unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/WebRequest/WebRequestController.cs

```diff
@@ -51,12 +53,164 @@ public static WebRequestController Create()
+        }
+
+        public async UniTask<UnityWebRequest> GetAudioClipAsync(
+            string url,
+            AudioType audioType,
+            Action<UnityWebRequest> onSuccess = null,
+            Action<UnityWebRequest> onfail = null,
+            int requestAttemps = 3,
+            int timeout = 0,
+            CancellationToken cancellationToken = default)
+        {
+            audioClipWebRequestFactory.SetAudioType(audioType);
+            return await SendWebRequest(audioClipWebRequestFactory, url, null, onSuccess, onfail, requestAttemps,
+                timeout, cancellationToken);
+        }
+        
+
+        private async UniTask<UnityWebRequest> SendWebRequest<T>(
+            T requestFactory,
+            string url,
+            DownloadHandler downloadHandler,
+            Action<UnityWebRequest> onSuccess,
+            Action<UnityWebRequest> onFail,
+            int requestAttemps,
+            int timeout,
+            CancellationToken cancellationToken,
+            Dictionary<string, string> headers = null) where T : IWebRequestFactory
+        {
+            requestAttemps = Mathf.Max(1, requestAttemps);
+            for (int i = requestAttemps - 1; i >= 0; i--)
+            {
+                UnityWebRequest request = requestFactory.CreateWebRequest(url);
+                request.timeout = timeout;
+
+                if (headers != null)
+                {
+                    foreach (var item in headers)
+                        request.SetRequestHeader(item.Key, item.Value);
+                }
+
+                if (downloadHandler != null)
+                    request.downloadHandler = downloadHandler;
+
+                UnityWebRequestAsyncOperation asyncOp = request.SendWebRequest();
+                while (!asyncOp.isDone)
+                {
+                    await UniTask.Yield();
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        request.Abort();
+                        break;
+                    }
+                }
+
+                if (asyncOp.webRequest == null)
+                    continue;
+
+                if (request.WebRequestSucceded())
+                {
+                    onSuccess?.Invoke(asyncOp.webRequest);
+                    return asyncOp.webRequest;
+                }
+                else if (!request.WebRequestAborted() && request.WebRequestServerError())
+                {
+                    if (requestAttemps > 0)
+                        continue;
+                    else
+                    {
+                        onFail?.Invoke(asyncOp.webRequest);
+                        return asyncOp.webRequest;
+                    }
+                }
+                else
+                {
+                    onFail?.Invoke(asyncOp.webRequest);
+                    return asyncOp.webRequest;
+                }
+            }
+
+            return null;
```

Comment:
`null` in the `async` flow does not make much sense. Imagine that consumers will need to check it for nullity every time.
Consider introducing an exception if attempts are exceeded

Reply:
This should be never reachable code, not a single edge case should be able to reach there, that's why I think throwing an exception is a good idea, something really bad happened if we reach it, will change it.